src/common/meta/src/key.rs

```diff
@@ -545,48 +577,125 @@ impl TableMetadataManager {
     pub async fn delete_table_metadata(
         &self,
         table_id: TableId,
         table_name: &TableName,
-        region_routes: &[RegionRoute],
+        table_route_value: &TableRouteValue,
     ) -> Result<()> {
-        // Deletes table name.
-        let table_name = TableNameKey::new(
+        // Builds keys
+        let datanode_ids = if table_route_value.is_physical() {
+            region_distribution(table_route_value.region_routes()?)
+                .into_keys()
+                .collect()
+        } else {
+            vec![]
+        };
+        let mut keys = Vec::with_capacity(3 + datanode_ids.len());
+        let table_name_key = TableNameKey::new(
             &table_name.catalog_name,
             &table_name.schema_name,
             &table_name.table_name,
         );
+        let table_info_key = TableInfoKey::new(table_id);
+        let table_route_key = TableRouteKey::new(table_id);
+        let datanode_table_keys = datanode_ids
+            .into_iter()
+            .map(|datanode_id| DatanodeTableKey::new(datanode_id, table_id))
+            .collect::<Vec<_>>();
+
+        keys.push(Key::atomic(table_name_key.as_raw_key()));
+        keys.push(Key::new(table_info_key.as_raw_key()));
+        keys.push(Key::new(table_route_key.as_raw_key()));
+        for key in &datanode_table_keys {
+            keys.push(Key::new(key.as_raw_key()));
+        }
+        self.tombstone_manager.create(keys).await?;
+        Ok(())
+    }
 
-        let delete_table_name_txn = self.table_name_manager().build_delete_txn(&table_name)?;
+    /// Restores metadata for table.
+    /// The caller MUST ensure it has the exclusive access to `TableNameKey`.
+    pub async fn restore_table_metadata(
+        &self,
+        table_id: TableId,
+        table_name: &TableName,
+        table_route_value: &TableRouteValue,
+    ) -> Result<()> {
+        // Builds keys
+        let datanode_ids = if table_route_value.is_physical() {
+            region_distribution(table_route_value.region_routes()?)
+                .into_keys()
+                .collect()
+        } else {
+            vec![]
+        };
+        let mut keys = Vec::with_capacity(3 + datanode_ids.len());
+        let table_name = TableNameKey::new(
+            &table_name.catalog_name,
+            &table_name.schema_name,
+            &table_name.table_name,
+        );
+        let table_info_key = TableInfoKey::new(table_id);
+        let table_route_key = TableRouteKey::new(table_id);
+        let datanode_table_keys = datanode_ids
+            .into_iter()
+            .map(|datanode_id| DatanodeTableKey::new(datanode_id, table_id))
+            .collect::<Vec<_>>();
+
+        keys.push(Key::atomic(table_name.as_raw_key()));
+        keys.push(Key::new(table_info_key.as_raw_key()));
+        keys.push(Key::new(table_route_key.as_raw_key()));
+        for key in &datanode_table_keys {
+            keys.push(Key::new(key.as_raw_key()));
+        }
```

Comment:
Perhaps we can add a "build_table_metadata_keys" to reuse this code, since I saw that the  `delete_table_metadata` also use it.